### ISSUE TITLE
Upgrading Ubuntu 18.04 to 20.04 or 22.04

### ISSUE THREAD

**Issue by infotts-py** [BookStackApp/BookStack#3930]
### Attempted Debugging

- [X] I have read the debugging page

### Searched GitHub Issues

- [X] I have searched GitHub for the issue.

### Describe the Scenario

When upgrading from Ubuntu 18.04 to 20.04 or 22.04 Bookstack is unresponsive after. I have no issues upgrading Ubuntu on other VM's. Upgraded Bookstack to 22.11.1 and all updates to Ubuntu 18.04 have been done prior to the major release upgrade.

Has anybody done this on a VM running Bookstack ? What should I look for to troubleshoot this issue after upgrading?

I have reverted back to a previous checkpoint at 18.04 and everything works.

### Exact BookStack Version

22.11.1

### PHP Version

8.0.26

### Hosting Environment

Ubuntu 20.04

**Comment by ssddanbrown** [comment]
This should be doable (Something I've commonly done) but sometimes things need to be tweaked for new versions of packages.

> What should I look for to troubleshoot this issue after upgrading?

Your webserver log would be a good start, it should show any errors assuming your webserver is running okay. Assuming apache is in use this is commonly found at ` /var/log/apache2/error.log`. 

My guess would be the issue is related to PHP package versions, and their usage with apache.
Sometimes you need to disable the old version of PHP then enable the new php in apache. For example, if going from PHP7.4 to 8.0:

```bash
sudo a2dismod php7.4
sudo a2enmod php8.0
```

The above is just a guess though.

**Comment by infotts-py** [comment]
PHP 7.2 is disabled and PHP 8.0.26 installed and enabled.. Repeated this after Ubuntu release upgrade again this morning. Composer is also at the latest version.

This all I've got in the apache error log :
[Tue Dec 27 09:41:06.493268 2022] [mpm_prefork:notice] [pid 798] AH00169: caught SIGTERM, shutting down
[Tue Dec 27 09:41:06.560102 2022] [mpm_prefork:notice] [pid 23811] AH00163: Apache/2.4.41 (Ubuntu) OpenSSL/1.1.1f configured -- resuming normal operations
[Tue Dec 27 09:41:06.560145 2022] [core:notice] [pid 23811] AH00094: Command line: '/usr/sbin/apache2'
[Tue Dec 27 09:41:38.354701 2022] [mpm_prefork:notice] [pid 23811] AH00171: Graceful restart requested, doing restart
[Tue Dec 27 09:41:38.378713 2022] [mpm_prefork:notice] [pid 23811] AH00163: Apache/2.4.41 (Ubuntu) OpenSSL/1.1.1f configured -- resuming normal operations
[Tue Dec 27 09:41:38.378723 2022] [core:notice] [pid 23811] AH00094: Command line: '/usr/sbin/apache2'
[Tue Dec 27 09:46:25.176330 2022] [mpm_prefork:notice] [pid 23811] AH00169: caught SIGTERM, shutting down
[Tue Dec 27 09:46:39.258250 2022] [mpm_prefork:notice] [pid 783] AH00163: Apache/2.4.41 (Ubuntu) OpenSSL/1.1.1f configured -- resuming normal operations
[Tue Dec 27 09:46:39.270320 2022] [core:notice] [pid 783] AH00094: Command line: '/usr/sbin/apache2'

I don't know where to go from here. Any ideas ?

**Comment by ssddanbrown** [comment]
- Are you sure you were using apache before and not nginx (Or another webserver)?
- Is apache staying up/alive (`sudo systemctl status apache2.service`). 
- Is MySQL staying up/alive (`systemctl status mysql.service`).
- If you watch the `/var/log/apache2/error.log` file do you see new entries appear when accessing the instance via the browser?
- Are there other error log files in the `/var/log/apache2/` directory?
- What exactly are you actually seeing in the browser? 
- Are you seeing anything in the bookstack error log on attempted access (`storage/logs/laravel.log` file within BookStack folder).

**Comment by infotts-py** [comment]
Are you sure you were using apache before and not nginx (Or another webserver)? Yes

Is apache staying up/alive (sudo systemctl status apache2.service). active (running)

Is MySQL staying up/alive (systemctl status mysql.service).  

```bash
● **mysql.service
     Loaded: masked (Reason: Unit mysql.service is masked.)
     Active: inactive (dead)**
```

I tried :

```bash
systemctl unmask mysql.service
service mysql start
systemctl status mysql.service
● mysql.service - LSB: Start and stop the mysql database server daemon
     Loaded: loaded (/etc/init.d/mysql; generated)
     Active: **active (exited)** since Thu 2022-12-29 11:47:57 EST; 12min ago
       Docs: man:systemd-sysv-generator(8)
    **Process: 2430 ExecStart=/etc/init.d/mysql start (code=exited, status=0/SUCCESS)**
)
```

If you watch the /var/log/apache2/error.log file do you see new entries appear when accessing the instance via the browser? No

Are there other error log files in the /var/log/apache2/ directory? yes but all empty

What exactly are you actually seeing in the browser? HTTP Error 500

Are you seeing anything in the bookstack error log on attempted access (storage/logs/laravel.log file within BookStack folder). :

```
Next Illuminate\\View\\ViewException: SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from `users` where `system_name` = public limit 1) (View: /var/www/bookstack/resources/views/layouts/base.blade.php) (View: /var/www/bookstack/resources/views/layouts/base.blade.php) in /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Database/Connection.php:712
Stack trace:
#0 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/View/Engines/PhpEngine.php(60): Illuminate\\View\\Engines\\CompilerEngine->handleViewException()
#1 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/View/Engines/CompilerEngine.php(61): Illuminate\\View\\Engines\\PhpEngine->evaluatePath()
#2 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/View/View.php(139): Illuminate\\View\\Engines\\CompilerEngine->get()
#3 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/View/View.php(122): Illuminate\\View\\View->getContents()
#4 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/View/View.php(91): Illuminate\\View\\View->renderContents()
#5 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Http/Response.php(69): Illuminate\\View\\View->render()
#6 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Http/Response.php(35): Illuminate\\Http\\Response->setContent()
#7 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Routing/ResponseFactory.php(55): Illuminate\\Http\\Response->__construct()
#8 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Routing/ResponseFactory.php(85): Illuminate\\Routing\\ResponseFactory->make()
#9 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(585): Illuminate\\Routing\\ResponseFactory->view()
#10 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(490): Illuminate\\Foundation\\Exceptions\\Handler->renderHttpException()
#11 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(356): Illuminate\\Foundation\\Exceptions\\Handler->prepareResponse()
#12 /var/www/bookstack/app/Exceptions/Handler.php(66): Illuminate\\Foundation\\Exceptions\\Handler->render()
#13 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php(196): BookStack\\Exceptions\\Handler->render()
#14 /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php(173): Illuminate\\Foundation\\Bootstrap\\HandleExceptions->renderHttpResponse()
#15 [internal function]: Illuminate\\Foundation\\Bootstrap\\HandleExceptions->handleException()
#16 {main}
  thrown at /var/www/bookstack/vendor/laravel/framework/src/Illuminate/Database/Connection.php:712)
[stacktrace]
#0 {main}
"}
```

So it seems to be an issue with mysql not starting yet I can't fix it. I've tried a ton of solutions found on internet, none of them worked for me. Any idea where to go from here ?

**Comment by ssddanbrown** [comment]
Ideally you'd need to identify exactly what is wrong with MySQL.
Check the MySQL error log for details (commonly at `/var/log/mysql/error.log`).

**Comment by infotts-py** [comment]
Yes, I looked at that log and it's empty.

**Comment by ssddanbrown** [comment]
Are there other log files in that folder with content?
What's the output of `ls -alh /var/log/mysql`?

**Comment by ssddanbrown** [comment]
Also worth watching journalctl for mysql while attempting to restart in another terminal:

```bash
journalctl -u mysql.service -f
```

**Comment by infotts-py** [comment]
/var/log/mysql$ ls -alh /var/log/mysql
total 36K
drwxr-x---  2 mysql adm    4.0K Dec 30 09:11 .
drwxrwxr-x 13 root  syslog 4.0K Dec 30 09:11 ..
-rw-r-----  1 mysql adm       0 Dec 30 09:11 error.log
-rw-r-----  1 mysql adm    2.5K Dec 27 09:31 error.log.1.gz
-rw-r-----  1 mysql adm      20 Dec 26 06:25 error.log.2.gz
-rw-r-----  1 mysql adm      20 Dec 25 06:25 error.log.3.gz
-rw-r-----  1 mysql adm      20 Dec 24 06:25 error.log.4.gz
-rw-r-----  1 mysql adm      20 Dec 23 06:25 error.log.5.gz
-rw-r-----  1 mysql adm    2.1K Dec 22 19:55 error.log.6.gz
-rw-r-----  1 mysql adm      20 Dec 21 06:25 error.log.7.gz

error.log.6.gz hasn't been modified. Upgrade done on the 27th only added to error.log.1.gz. (I've been going back and forth with checkpoints from before and after upgrade to 20.04). No new error logs when I unmask and try tpo start the mysql service.

docs@docs:/var/log/mysql$ vi error.log.1.gz
2022-12-27T14:15:11.320220Z 0 [Note] Giving 0 client threads a chance to die gracefully
2022-12-27T14:15:11.320256Z 0 [Note] Shutting down slave threads
2022-12-27T14:15:11.320261Z 0 [Note] Forcefully disconnecting 0 remaining clients
2022-12-27T14:15:11.320267Z 0 [Note] Event Scheduler: Purging the queue. 0 events
2022-12-27T14:15:11.357191Z 0 [Note] Binlog end
2022-12-27T14:15:11.359627Z 0 [Note] Shutting down plugin 'auth_socket'
2022-12-27T14:15:11.363828Z 0 [Note] Shutting down plugin 'ngram'
2022-12-27T14:15:11.363837Z 0 [Note] Shutting down plugin 'partition'
2022-12-27T14:15:11.363841Z 0 [Note] Shutting down plugin 'BLACKHOLE'
2022-12-27T14:15:11.363845Z 0 [Note] Shutting down plugin 'ARCHIVE'
2022-12-27T14:15:11.363848Z 0 [Note] Shutting down plugin 'PERFORMANCE_SCHEMA'
2022-12-27T14:15:11.363878Z 0 [Note] Shutting down plugin 'MRG_MYISAM'
2022-12-27T14:15:11.363881Z 0 [Note] Shutting down plugin 'MyISAM'
2022-12-27T14:15:11.363888Z 0 [Note] Shutting down plugin 'INNODB_SYS_VIRTUAL'
2022-12-27T14:15:11.363891Z 0 [Note] Shutting down plugin 'INNODB_SYS_DATAFILES'
2022-12-27T14:15:11.363894Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESPACES'
2022-12-27T14:15:11.363896Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN_COLS'
2022-12-27T14:15:11.363899Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN'
2022-12-27T14:15:11.363901Z 0 [Note] Shutting down plugin 'INNODB_SYS_FIELDS'
2022-12-27T14:15:11.363904Z 0 [Note] Shutting down plugin 'INNODB_SYS_COLUMNS'
2022-12-27T14:15:11.363906Z 0 [Note] Shutting down plugin 'INNODB_SYS_INDEXES'
2022-12-27T14:15:11.363909Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESTATS'
2022-12-27T14:15:11.363911Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLES'
2022-12-27T14:15:11.363914Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_TABLE'
2022-12-27T14:15:11.363916Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_CACHE'
2022-12-27T14:15:11.363919Z 0 [Note] Shutting down plugin 'INNODB_FT_CONFIG'
2022-12-27T14:15:11.363921Z 0 [Note] Shutting down plugin 'INNODB_FT_BEING_DELETED'
2022-12-27T14:15:11.363924Z 0 [Note] Shutting down plugin 'INNODB_FT_DELETED'
2022-12-27T14:15:11.363927Z 0 [Note] Shutting down plugin 'INNODB_FT_DEFAULT_STOPWORD'
2022-12-27T14:15:11.363929Z 0 [Note] Shutting down plugin 'INNODB_METRICS'
2022-12-27T14:15:11.363932Z 0 [Note] Shutting down plugin 'INNODB_TEMP_TABLE_INFO'
2022-12-27T14:15:11.363935Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_POOL_STATS'
2022-12-27T14:15:11.363938Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE_LRU'
2022-12-27T14:15:11.363941Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE'
2022-12-27T14:15:11.363943Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX_RESET'
2022-12-27T14:15:11.363946Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX'
2022-12-27T14:15:11.363948Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM_RESET'
2022-12-27T14:15:11.363951Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM'
2022-12-27T14:15:11.363954Z 0 [Note] Shutting down plugin 'INNODB_CMP_RESET'
2022-12-27T14:15:11.363956Z 0 [Note] Shutting down plugin 'INNODB_CMP'
2022-12-27T14:15:11.363959Z 0 [Note] Shutting down plugin 'INNODB_LOCK_WAITS'
2022-12-27T14:15:11.363961Z 0 [Note] Shutting down plugin 'INNODB_LOCKS'
2022-12-27T14:15:11.363964Z 0 [Note] Shutting down plugin 'INNODB_TRX'
2022-12-27T14:15:11.363967Z 0 [Note] Shutting down plugin 'InnoDB'
2022-12-27T14:15:11.364007Z 0 [Note] InnoDB: FTS optimize thread exiting.
2022-12-27T14:15:11.364061Z 0 [Note] InnoDB: Starting shutdown...
2022-12-27T14:15:11.464228Z 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
2022-12-27T14:15:11.464959Z 0 [Note] InnoDB: Buffer pool(s) dump completed at 221227  9:15:11
2022-12-27T14:15:12.773219Z 0 [Note] InnoDB: Shutdown completed; log sequence number 4996157105
2022-12-27T14:15:12.774230Z 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2022-12-27T14:15:12.774242Z 0 [Note] Shutting down plugin 'MEMORY'
2022-12-27T14:15:12.774246Z 0 [Note] Shutting down plugin 'CSV'
2022-12-27T14:15:12.774250Z 0 [Note] Shutting down plugin 'sha256_password'
2022-12-27T14:15:12.774259Z 0 [Note] Shutting down plugin 'mysql_native_password'
2022-12-27T14:15:12.774395Z 0 [Note] Shutting down plugin 'binlog'
2022-12-27T14:15:12.774633Z 0 [Note] /usr/sbin/mysqld: Shutdown complete

2022-12-27T14:15:35.451512Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2022-12-27T14:15:35.452863Z 0 [Note] /usr/sbin/mysqld (mysqld 5.7.40-0ubuntu0.18.04.1) starting as process 1530 ...
2022-12-27T14:15:35.498925Z 0 [Note] InnoDB: PUNCH HOLE support available
2022-12-27T14:15:35.498943Z 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2022-12-27T14:15:35.498946Z 0 [Note] InnoDB: Uses event mutexes
"error.log.1.gz" [readonly][noeol] 8L, 2541C                                                                                                                                                                               1,1           Top


journalctl -u mysql.service -f
Dec 30 09:12:16 docs systemd[1]: Started LSB: Start and stop the mysql database server daemon.

**Comment by ssddanbrown** [comment]
The MySQL version in your error log (`mysqld 5.7.40-0ubuntu0.18.04.1`) looks to be for Ubuntu 18.04. Maybe it's not being upgraded during the upgrade process to remain compatible. From Ubuntu 20.04 MySQL 8.0 is used. 
Might have been held back potentially related to the service being masked. 

Assuming you have a safe backup of a previous instance, I'd try a re-install of MySQL on the upgraded system, ensuring the correct version is being used. (Currently `8.0.31-0ubuntu0.20.04.1` for Ubuntu 20.04 or `8.0.31-0ubuntu0.22.04.1` for Ubuntu 22.04).

**Comment by infotts-py** [comment]
I was on vacation last week but got back to this tonight. I wasn't able to get MySQL updated correctly from 5.7 to 8.031. I ended up going back to my checkpoint and starting over from scratch.

- Upgraded Ubuntu 18.04 to 20.04
- MySQL 5.7 was removed during upgrade process but no other version installed. Database was kept.
- After upgrade & reboot MySQL wasn't installed so installed the latest version 8.031.
- Everything seems to work.
- Upgraded Ubuntu 20.04 to 22.04
- Everything seems to work!

I don't know why the upgrade process didn't react the same way the previous times... 

Thank you for your help!
